README
                  Sanewall - making sense of firewalling
                  **************************************
                       http://www.sanewall.org/
                       ************************

Sanewall is a firewall builder for Linux which uses an elegant language,
abstracted to just the right level. This makes it powerful as well as easy
to use, audit, and understand. It allows you to create very readable
configurations even for complex stateful firewalls.

Simple but useful firewalls need only a few lines of configuration. Very
complex setups with flow control and external commands can be created by
using bash (http://www.gnu.org/software/bash/) commands in-line with the
standard configuration language.

Sanewall can be used to produce local firewalls or router-based firewalls
on any system that bash and iptables are available including full GNU/Linux
servers and embedded routers such as OpenWRT (http://www.openwrt.org/).


Known Issues
============
This is a development release of Sanewall. Below is a list of issues you
should consider before using it. Note that the list is probably not
exhaustive.

You should also read the "IPv6 Support" section of this README.

If you do choose to use this version, please take the time to audit
the output and ensure it meets your expectations.

If you encounter any bugs or unexpected behaviour please report them
to the mailing list or bug tracker.

1. There are a variety of ICMPv6 packet types without helpers and
   not described here yet which are needed for full handling of IPv6.

   The ipv6error packet rules have not been thoroughly tested.

   A full set of ICMPv6 recommendations that ought to be implemented are
   here:
     http://tools.ietf.org/html/rfc4890

   The recommendations, and how to achieve them (and where there is
   still work to do) are documented in the "RFC 4890 Recommendations"
   section of sanewall-manual.pdf and the online documentation.

2. There is no Linux kernel NAT support for IPv6, you will get errors
   if you try to use it.

   The masquerade command automatically defaults to  only.

3. No way to successfully mark a whole interface or router as ipv4/ipv6 only

   There hopefully aren't too many situations where one would want
   to do this but support will be added in due course.

4. Private / Unroutable IPv6 lists may be incomplete.

5. Creating a rule which applies only to one ip type will lead to
   empty chains. For example:
      server ssh accept src "10.0.0.0/8"

   will create two chains, one using iptables and the other using
   ip6tables; only the iptables chain has any rules added to it. This
   should not cause any problems, but can look odd when inspecting
   the tables directly.


Goals
=====
The main goals of the Sanewall project are:
  - Allow experts and non-experts alike to produce secure firewalls
  - Simple configuration should be simple
  - Complex configuration should be possible (and as easy as possible)
  - Maintain compatibility with old FireHOL configurations
  - Eliminate distinctions between IPv4 and IPv6 wherever possible
  - Keep to a minimal set of dependencies

This is achieved by providing an expressive, easy-to-read, write and
understand configuration language geared specifically to firewalls
(a Domain Specific Language).

The language is sufficiently brief, well-structured and meaningful that
a human can manage the firewall rules without the need for additional tools.
The use of a simple text-file for configuration allows for optimal use
with version control and file-comparison tools.


Getting Started
===============
If you want to install the package from the source tar-files found here:
   http://download.sanewall.org/releases
please read the file INSTALL first. Sanewall uses the GNU Autotools so
you can get away with:
  ./configure && make && make install

When you first install the program a very basic sanewall.conf.example
is installed, which if you rename it to sanewall.conf will allow connections
out but not in. To get something more complete you have three choices:

  1. If you are replacing FireHOL you should just be able to create the
     Sanewall configuration by copying firehol.conf over sanewall.conf
     and renaming any FIREHOL_ variables to SANEWALL_ instead.

     Rename the files (if you have them, they are optional):
        RESERVED_IPS   ->   RESERVED_IPV4
        PRIVATE_IPS    ->   PRIVATE_IPV4
        MULTICAST_IPS  ->   MULTICAST_IPV4

     There are new (optional) files for IPv6 addresses:
        RESERVED_IPV6, PRIVATE_IPV6 and MULTICAST_IPV6

     If you have any custom services in /etc/firehol/services then you
     will need to update those, too.

     If you made use of "firehol save" you need to verify that the
     SANEWALL_AUTOSAVE and SANEWALL_AUTOSAVE6 variables are correctly
     specified.

     If you made use of "firehol condrestart" you should be aware that
     sanewall implements the behaviour documented here:
       http://fedoraproject.org/wiki/Packaging:SysVInitScript#Required_Actions
     whereas firehol started only if it was not already running.

  2. Start from an example configuration:
       client-all.conf
       lan-gateway.conf
       server-dmz.conf
       office.conf
       ipv6.conf

  3. Have Sanewall try to generate a configuration tailored to the machine
     automatically by running:
       sanewall wizard

You should review the variables that can be configured and decide if you
want to change any. The variables are documented in the "control variables"
reference section of sanewall-manual.pdf and the online documentation. You
can also read the man-page:
  man sanewall-variables

If you are running a service which is not pre-defined for you it is simple
to define your own. This is documented in the "Adding Services" part of
the "sanewall configuration" reference section of sanewall-manual.pdf and
the online documentation. You can also read the man-page:
  man sanewall.conf

Finally, you will also want to ensure that Sanewall runs at boot-time. If you
installed from an official package this will be configured in the usual way.
For a source installation, the binary can be linked directly into /etc/init.d
on many systems. In addition, some example init scripts are available here:
  http://bugs.sanewall.org/wiki/Init_Scripts


IPv6 support
============
Versions of Sanewall above 1.1.0 contain IPv6 support.

Sanewall will attempt to load both iptables and ip6tables entries for
all rules, except as follows:

1. When invoked with -4 or -6 on the command line:

     sanewall -4 ...
       IPv4 only

     sanewall -6 ...
       IPv6 only

2. Explicitly stated version in the configuration e.g.:

   interface wlan0 internet
      ipv4 masquerade
      ipv6 server ssh allow

   Will only enable masquerading on IPv4 and ssh access via IPv6.

2. Implicit protocol or address in the configuration e.g.:

   interface eth0 lan
      server telnet accept src "192.168.0.0/24"
      server ftp accept src "fe80::/64"
      server ssh accept src "fe80::/64 192.168.0.0/24"

   Will should do the "right thing" and allow mix and match of addresses.
   Note that creating a rule with e.g. "src not" will negate fully. For
   example:
      server ssh accept src not "10.0.0.0/8"

   Will accept any address that is not 10.0.0.0/8, including any IPv6.
   If you mean not 10.0.0.0/8 and no IPv6 addresses you should restrict
   the rule explicitly:
      ipv4 server ssh accept src not "10.0.0.0/8"

   The same applies if you negate only IPv6 addresses.

3. ICMP and ICMPv6 are not identical. You will get warnings and/or errors
   if you try to include rules for an ICMP packet type for ICMPv6 and
   vice-versa e.g.:

   interface eth0 lan
      client timestamp accept

   will issue a warning, but:

   interface eth0 lan
      ipv4 timestamp client accept

   will not, the reason being that the ICMP packets timestamp-request and
   timestamp-reply have no equivalent in ICMPv6.

   New ipv6-only marked protocols (e.g. ipv6router, ipv6neigh) will not
   warn in "both" mode, only in "ipv4" mode.

4. Network neighbour solicitation/advertisement messages (IPv6 replacement
   for ARP) should be enabled on each interface. In most contexts,
   router solicitation/advertisement messages should also be enabled
   to allow router auto-configuration.

   See the documentation (sanewall-manual.pdf or man sanewall-services)
   on ipv6neigh and ipv6router for details. See also the ipv6.conf example.

5. Various ICMPv6 error messages need to be allowed for correct operation
   of IPv6.

   See the documentation (sanewall-manual.pdf or man sanewall-services)
   on the ipv6error complex rule for details. See also the ipv6.conf example.


Support and documentation
=========================
The main website is here:
   http://www.sanewall.org/

To ask questions please sign up to the list:
   http://lists.sanewall.org/mailman/listinfo/sanewall-users

Man pages, PDF and HTML documentation are provided as part of the package
and can be found in the tarball or in your distribution's standard locations
(e.g. /usr/share/doc). The latest manual is always available as PDF and
online HTML here:
   http://download.sanewall.org/releases/latest/sanewall-manual.pdf
   http://download.sanewall.org/releases/latest/sanewall-manual.html

along with a list of all services supported "out of the box":
   http://download.sanewall.org/releases/latest/sanewall-services.html

For further help and advice the sanewall-users mailing list archive is
fully searchable:
   http://lists.sanewall.org/pipermail/sanewall-users/


Contributing
============
The wiki page for contributors and potential contributors is here:
   http://bugs.sanewall.org/wiki/Getting_Involved

The official bug tracker is here:
   http://bugs.sanewall.org/sanewall

The official git trees are here:
   http://git.sanewall.org/

If you would like to get involved, please consider subscribing to the
development mailing list:
   http://lists.sanewall.org/mailman/listinfo/sanewall-dev


History
=======
Sanewall is a fork of FireHOL (http://firehol.sourceforge.net/) which was
made when development of that project stalled. A great deal is owed to that
project and Costa Tsaousis for originating it. All existing FireHOL
definitions should be compatible with Sanewall.


License
=======
Copyright (C) 2012,2013 Phil Whineray <phil@sanewall.org>
Copyright (C) 2003-2013 Costa Tsaousis <costa@tsaousis.gr>

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA


NEWS
2013/08/19:

Release of version 1.1.5

Uses flock(1) instead of lockfile(1) since util-linux should be more
commonly installed than procmail.

Fixes IPv4/IPv6 detection for older versions of iptables(8). The "-C" flag
is not always present.

Cleanups and minor improvements:
 - Improve error message when an expected command not found
 - Stop emitting duplicate rules during a "panic"
 - Remove the blank line from wizard/helpme output
 - Remove the #! from wizard output - it cannot be executed directly
 - Man-page for "explain" firehol now notes the output cannot be
   used with iptables(8) verbatim
 - Regression tests should now work with any iptables version
 - Point to the mailing list for announcements
 - Cleaned up boilerplace information in script

2013/07/06:

Release of version 1.1.4

"firehol save" now creates two files using FIREHOL_AUTOSAVE (IPv4)
and FIREHOL_AUTOSAVE6 (IPv6).

The default values for these variables are now detected at configure
time rather than run-time. Most command detection is now also at
configure-time rather than runtime.

The autodetection should work on most common platforms provided the
appropriate packages making use of the saved rules are installed (e.g.
iptables-persistent on Debian).

Maintainers in particular may prefer to set the values explicitly:
  --with-autosave=PATH
  --with-autosave6=PATH

If the documentation is being rebuilt using --enable-maintainer-mode
it will be updated to reflect the default path.

If no default is found or the --without-zzz form is configured, the user
will have to declare the variables themselves to make use of the save
function.

"firehol status" now identifies which blocks are IPv6 or IPv4.

"firehol condrestart" now follows conventions, i.e. restarting the
firewall if it appears to already be running, rather than the old
behaviour (starting it only if not run before).


2013/05/12:

Release of version 1.1.3

Fix IPv4/IPv6 auto-detection so that is does not get confused by VLAN
interface names such as eth0.22.

Reduced the autotools version requirement to 2.63 since this is reported
as working fine.


2013/05/12:

Release of version 1.1.2

Fix kernel version detection so that it is more flexible and less error
prone (3.8-1-amd64 was breaking the original).

Configure script makes firehol executable. The unconfigured firehol.in
issues a warning when it is run directly.

Configure script now sets /usr/local/etc as the location for FireHOL
to look in as well as store configuration files in if --sysconfdir is not
given.

Allow switching on of debug output by setting environment variable
FIREHOL_DEBUGGING to a non-empty value.

Handle domain names that refer to records that are IPv4, IPv6 or both
(previously they were assumed to be IPv4 only).

Fix protection against direct use of /sbin/iptables and /sbin/ip6tables
that were broken from 1.1.0.

Note that the fixes to IPv4/IPv6 matching from version 1.1.1 also had
the effect of resolving the problems previosly associated with "src not"
on all rules and "src" on some. Both of these constructs should now be
fine to use without an ipv4/ipv6 prefix. Note however that:
   server ssh accept src not some-IPv4

Accepts all non-matching IPv4 addresses and all IPv6 addresses (since none
match) so if you want to stop ipv6 addresses you need to do so by limiting
the rule to IPv4 space:
   ipv4 server ssh accept src not some-IPv4


2013/05/06:

Release of version 1.1.1

When startup fails, both IPv4 and IPv6 firewalls are correctly restored.

Fixes "mac" helper command for versions of iptables 1.4.12+ as per 1.0.1
and also prevent mac addresses being seen as IPv6 addresses.

Added a regression test framework.

Improve IPv4/IPv6 address matching (some internal names such as mark.1 were
being identified as IPv4 only).


2013/04/24:

Development series 1.1.0 starts

Adds IPv6 support

Can be invoked as:
  firehol [whatever]     -- "Both" mode (default)
  firehol -4 [whatever]  -- IPv4 mode (should show original behaviour)
  firehol -6 [whatever]  -- IPv6 mode (ip6tables only)

Please make sure you read the "Known Issues" and "IPv6 Support" sections
of the README. If you choose to use this development version, please take
the time to audit the output and ensure it meets your expectations and
report any bugs or unexpected behaviour to the mailing list or bug tracker.

Rename these configuration files (if you have them, they are optional):
  RESERVED_IPS   ->   RESERVED_IPV4
  PRIVATE_IPS    ->   PRIVATE_IPV4
  MULTICAST_IPS  ->   MULTICAST_IPV4

There are new (optional) files for IPv6 addresses:
  RESERVED_IPV6, PRIVATE_IPV6 and MULTICAST_IPV6


2013/04/24:

Release of version 1.0.0

No functional changes from 0.9.10.

Slight changes to README, license snippet in script brought in line with
recommended wording.


2013/04/13:

Release of version 0.9.10

This version allows certain variables to be set via the environment as
well as via the firehol.conf file.

See the PDF/HTML manual or firehol-variables manpage for details.


2013/04/06:

Release of version 0.9.9

This version now uses GNU autotools so you can run:
  ./configure && make && make install

in the usual way. Note that the default is to install under /usr/local,
so if you have previously installed from source, you may want to run
configure more like this instead:
  ./configure --prefix=/usr --sysconfdir=/etc

Read the INSTALL file and ./configure --help for more information.


2013/03/23:

Release of version 0.9.8


mkservicelist.pl
#!/usr/bin/perl -w

my $debug = 0;

use strict;
use Data::Dumper;
use Text::ParseWords;

#
# Simple line-oriented parser helper routines
#

my $filename;
my $fh;
my $line;
my $lineno;

sub getpos {
  return "$filename:$lineno";
}

sub set {
  $filename = shift;
  $fh = shift;
  $lineno = shift;
  $line = undef;
}

sub curr_line {
  return $line;
}

sub next_line_int {
  $line = <$fh>;
  $lineno++;
  unless (defined($line)) {
    close $fh;
    return;
  }

  while ($line =~ /^[[:space:]]*#/) {
    $line = <$fh>;
    $lineno++;
    return unless defined($line);
  }

  chomp $line;
  $line =~ s/^[[:space:]]*#.*//;
  $line =~ s/[[:space:]]+$//;
}

sub next_line {
  next_line_int();
  while (defined($line) and $line =~ /^$/) {
    next_line_int();
  }
}

sub err {
  my $message = shift;
  die getpos() . ": $message\n";
}

sub e_exp {
  my $expected = shift;
  my $actual = shift;
  $actual = "<EOF>" unless defined($actual);
  err("expected '$expected', got '$actual'");
}

sub open_file {
  my $name = shift;
  die "No name given\n" unless $name;

  open my $fh, "<$name" or die "$name: Unable to open\n";
  return $fh;
}

if (@ARGV != 3) {
  print STDERR "Usage: mkservicedoc.pl output.xml firehol-script services-data.txt\n";
  exit 1;
}

$fh = open_file($ARGV[1]);

# Extract service information from script
my %services = ();
my %all_run = ();

while (<$fh>) {
  if (/^(server|client)_([[:alnum:]_]+)_ports="?([^"]*)"?/) {
    $services{$2}{$1} = $3;
  }
  if (/^(helper)_([[:alnum:]_]+)="?([^"]*)"?/) {
    my $field = $1;
    my $name = $2;
    my $val = $3;
    $services{$name}{$field} = $val;
    $services{$name}{"mod"} = $val;
    $services{$name}{"mod"} =~ s/([^[:space:]]+)/nf_conntrack_$1/g;
    $services{$name}{"mod"} =~ s/[[:space:]]+/,/g;
    $services{$name}{"modnat"} = $val;
    $services{$name}{"modnat"} =~ s/([^[:space:]]+)/nf_nat_$1/g;
    $services{$name}{"modnat"} =~ s/[[:space:]]+/,/g;
  }
  if (/^rules_([[:alnum:]_]+)\(\) {/) {
    $services{$1}{type} = "complex";
  }
  if (/^ALL_SHOULD_ALSO_RUN="/) {
    s/^[^ ]* //;
    s/"$//;
    for my $s (split(/[[:space:]]+/, $_)) {
      $all_run{$s} = 1;
    }
  }
}
close $fh;

my %single_cmd = (
                  'CPORT' => 'client',
                  'SPORT' => 'server',
                  'MOD' => 'mod',
                  'MODNAT' => 'modnat',
                  'NAME' => 'name',
                  'ALIAS' => 'alias',
                  'WIKI' => 'wiki',
                  'HOME' => 'home',
                 );

my %multi_cmd = (
                  'EXAMPLE' => 'example',
                  'NOTES' => 'notes',
                 );

# Extract service information from database
$filename = $ARGV[2];
$fh = open_file($ARGV[2]);
my %db = ();
my %dbalias = ();
next_line;
while (curr_line()) {
  my $name;
  if ($line =~ /^SERVICE (.*)/) {
    my $name = $1;
    next_line;
    last unless $line;
    while ($line =~ /^\t([[:alnum:]]+)/) {
      my $cmd = $1;
      $line =~ s/^[[:space:]]+[[:alnum:]]+[[:space:]]+//;
      if ($single_cmd{$cmd}) {
        my $n = $single_cmd{$cmd};
        $db{$name}{$n} = $line;
        if ($n eq "alias") {
          $dbalias{$line} = $name;
        }
        next_line;
        last unless $line;
      } elsif ($multi_cmd{$cmd}) {
        my $n = $multi_cmd{$cmd};
        my $val;
        next_line;
        last unless $line;
        while ($line =~ /^\t\t/) {
          $line =~ s/^[[:space:]]+//;
          $line =~ s/^-$//;
          if (defined($val)) { $val .= "\n" . $line } else { $val = $line; }
          next_line;
          last unless $line;
        }
        $db{$name}{$n} = $val if (defined($val));
      } else {
        err "Unknown command $cmd\n";
      }
      last unless $line;
    }
  } else {
    err "Unknown line $line\n";
  }
}
close $fh;

my $err;
my %servs = ();
foreach my $s (sort(keys(%services))) {
  unless ($db{$s} or $dbalias{$s}) {
    print STDERR "Service $s in firehol but not in services-data.txt\n";
    $err = 1;
  }
  $servs{$s} = 1;
}

foreach my $d (sort(keys(%db))) {
  unless ($services{$d}) {
    print STDERR "Service $d services-data.txt but not in firehol\n";
    $err = 1;
  }
  $servs{$d} = 1;
}

foreach my $k (sort(keys(%servs))) {
  my $name = $db{$k}{"name"};
  if (!defined($name) and !defined($dbalias{$k})) {
    print STDERR "Service $k has no NAME or ALIAS in services-db.txt\n";
    $err = 1;
  }
  if (defined($db{$k}{"mod"}) and !defined($services{$k}{"mod"})) {
    print STDERR "Service $k has services-db.txt MOD but no helper in script!\n";
    $err = 1;
  }
  if (defined($db{$k}{"modnat"}) and !defined($services{$k}{"modnat"})) {
    print STDERR "Service $k has services-db.txt MODNAT but no helper in script!\n";
    $err = 1;
  }
  if (defined($db{$k}{"mod"}) and defined($services{$k}{"mod"})) {
    if ($db{$k}{"mod"} eq $services{$k}{"mod"}) {
      print STDERR "Service $k MOD in services-db.txt is redundant\n";
      $err = 1;
    } elsif ($db{$k}{"mod"} ne "N/A" and $db{$k}{"mod"} !~ /^See/) {
      print STDERR "Service $k services-db.txt MOD, different to helper in script and not N/A!\n";
      $err = 1;
    }
  }
  if (defined($db{$k}{"modnat"}) and defined($services{$k}{"modnat"})) {
    if ($db{$k}{"modnat"} eq $services{$k}{"modnat"}) {
      print STDERR "Service $k MODNAT in services-db.txt is redundant\n";
      $err = 1;
    } elsif ($db{$k}{"modnat"} ne "N/A" and $db{$k}{"modnat"} !~ /^See/) {
      print STDERR "Service $k services-db.txt MODNAT, different to helper in script and not N/A!\n";
      $err = 1;
    }
  }
}
exit 1 if $err;

sub coalesce {
  while (@_ > 0) {
    $_ = shift @_;
    return $_ if defined $_;
  }
  return undef;
}

sub docmod {
  my $b = "http://cateee.net/lkddb/web-lkddb";
  my $modlist = shift @_;

  my @mods = split(/,/, $modlist);
  my @modwithurl = ();
  foreach my $mod (@mods) {
    if ($mod =~ /^nf/) {
      my $m = uc($mod);
      $m =~ s/CONNTRACK_PROTO/CT_PROTO/;
      push @modwithurl, "$mod (<ulink url=\"$b/$m.html\">CONFIG_$m</ulink>)";
    } else {
      push @modwithurl, $mod;
    }
  }

  $modlist = join(", ", @modwithurl);
  return $modlist;
}

open O, ">$ARGV[0]" or die;

print O '<!DOCTYPE refsect1 PUBLIC "-//OASIS//DTD DocBook XML V4.1.2//EN"' . "\n";
print O '                          "docbook/docbook-xml/docbookx.dtd">' . "\n";
print O "<refsect1><title>Services</title>\n";
my @keys = sort { return uc($a) cmp uc($b) } keys(%servs);
foreach my $k (@keys) {
  my $name = $db{$k}{"name"};
  my @links = ();
  my $home = $db{$k}{"home"};
  my $wiki = $db{$k}{"wiki"};
  push @links, "<ulink url=\"$home\">Homepage</ulink>" if $home;
  push @links, "<ulink url=\"$wiki\">Wikipedia</ulink>" if $wiki;

  my $sport = coalesce($db{$k}{"server"}, $services{$k}{"server"}, "N/A");
  my $cport = coalesce($db{$k}{"client"}, $services{$k}{"client"}, "N/A");
  my $netmod = coalesce($db{$k}{"mod"}, $services{$k}{"mod"});
  my $natmod = coalesce($db{$k}{"modnat"}, $services{$k}{"modnat"});
  my $example = coalesce($db{$k}{"example"}, $services{$k}{"example"});
  my $notes = coalesce($db{$k}{"notes"}, $services{$k}{"notes"});
  print O "\n";
  if ($dbalias{$k}) {
    my $official = $dbalias{$k};
    $name = $db{$official}{"name"};
  }

  print O "  <variablelist id=\"service-$k\">\n";
  print O "    <title>$k - $name\n";
  print O "      <indexterm>\n";
  print O "        <primary>$k</primary>\n";
  print O "        <secondary>$name</secondary>\n";
  print O "      </indexterm>\n";
  print O "    </title>\n";

  if ($dbalias{$k}) {
    print O "    <varlistentry><term>Alias</term>\n";
    print O "      <listitem>\n";
    print O "        <para>See <xref linkend=\"service-$dbalias{$k}\"/></para>\n";
    print O "      </listitem>\n";
    print O "    </varlistentry>\n";
    print O "  </variablelist>\n";
    next;
  }


  if ($example) {
    print O "    <varlistentry><term>Example</term>\n";
    print O "      <listitem><para>\n";
    print O "        <programlisting>$example</programlisting>\n";
    print O "      </para></listitem>\n";
    print O "    </varlistentry>\n";
  }
  print O "    <varlistentry><term>Server Ports</term>\n";
  print O "      <listitem><para>$sport</para></listitem>\n";
  print O "    </varlistentry>\n";
  print O "    <varlistentry><term>Client Ports</term>\n";
  print O "      <listitem><para>$cport</para></listitem>\n";
  print O "    </varlistentry>\n";
  if ($netmod) {
    print O "    <varlistentry><term>Netfilter Modules</term>\n";
    print O "      <listitem><para>" .docmod($netmod). "</para></listitem>\n";
    print O "    </varlistentry>\n";
  }
  if ($natmod) {
    print O "    <varlistentry><term>Netfilter NAT Modules</term>\n";
    print O "      <listitem><para>" .docmod($natmod). "</para></listitem>\n";
    print O "    </varlistentry>\n";
  }
  if (@links) {
    print O "    <varlistentry><term>Links</term>\n";
    print O "      <listitem><para>\n";
    print O "        " . join(", ", @links) . "\n";
    print O "      </para></listitem>\n";
    print O "    </varlistentry>\n";
  }
  if ($k eq "all") {
    my $xtra_notes = "The following complex services are activated:\n";
    $xtra_notes .= "<simplelist>\n";
    foreach my $s (sort(keys(%all_run))) {
      $xtra_notes .= "<member><xref linkend=\"service-$s\"/></member>\n";
    }
    $xtra_notes .= "</simplelist>\n";
    if (defined($notes)) {
      $notes .= "\n" . $xtra_notes;
    } else {
      $notes = $xtra_notes;
    }
  }
  if ($notes) {
    print O "    <varlistentry><term>Notes</term>\n";
    print O "      <listitem>\n";
    print O "        <para>\n";
    $notes =~ s/&/&amp;/g;
    $notes =~ s/&amp;nbsp;/&nbsp;/g;
    my @lines = split /\n/, $notes;
    my @paras = ();
    while (@lines > 0) {
      my $_ = shift @lines;
      if (@paras == 0) {
        push @paras, $_;
      } elsif ($_ eq "") {
        push @paras, "";
      } else {
        if ($paras[$#paras] eq "") {
          $paras[$#paras] = $_;
        } else {
          $paras[$#paras] .= "\n          $_";
        }
      }
    }
    print O "          " .
        join("\n        </para>\n        <para>\n          ", @paras) . "\n";
    print O "        </para>\n";
    print O "      </listitem>\n";
    print O "    </varlistentry>\n";
  }
  print O "  </variablelist>\n";
}
print O "\n</refsect1>\n";

#print STDERR Dumper(\%services) . "\n";
#print STDERR Dumper(\%db) . "\n";


examples/adblock.sh
#!/bin/bash -e
#
# $Id: adblock.sh,v 1.1 2003/10/07 23:42:17 ktsaou Exp $
# A script that will fetch the IPs of popular add servers.
#  - Updated to run FireHOL on success
#
# To use this, just put in your cron jobs, like this:
#
#      0 * * * * /path/to/adblock.sh >/etc/firehol/adblock-ips
#
# and then in your firehol.conf:
#
#     source /etc/firehol/adblock-ips
#
# later in the config file you can use (for example)
#
#     client http accept dst not "${ADSERVERS_IPS}"
#
# or even:
#
#     blacklist full "${ADSERVERS_IPS}"
#
PATH=/usr/local/sbin:/usr/local/bin:/sbin:/usr/sbin:/bin:/usr/bin
export PATH

printf "ADSERVERS_IPS=\""
printf "%q " `wget -q -O - "http://pgl.yoyo.org/adservers/iplist.php?ipformat=plain&showintro=0" | egrep -e "^[0-9]+\.[0-9]+\.[0-9]+\.[0-9]+\$"`
printf "\"\n"

firehol start


admin/get-iana.sh
#!/bin/bash

# $Id: get-iana.sh,v 1.15 2013/01/06 23:49:08 ktsaou Exp $
#
# $Log: get-iana.sh,v $
# Revision 1.15  2013/01/06 23:49:08  ktsaou
# Removed depedency to get-iana.sh
# It is not usefull any more.
#
# Revision 1.14  2010/06/07 15:44:09  ktsaou
# Made get-iana.sh support the latest IANA format.
#
# Revision 1.13  2010/04/08 22:03:08  ktsaou
# Removed --proxy=off for wget.
#
# Revision 1.12  2008/03/17 22:08:43  ktsaou
# Updated for latest IANA reservations format.
#
# Revision 1.11  2007/06/13 14:40:04  ktsaou
# *** empty log message ***
#
# Revision 1.10  2007/05/05 23:38:31  ktsaou
# Added support for external definitions of:
#
# RESERVED_IPS
# PRIVATE_IPS
# MULTICAST_IPS
# UNROUTABLE_IPS
#
# in files under the same name in /etc/firehol/.
# Only RESERVED_IPS is mandatory (firehol will complain if it is not there,
# but it will still work without it), and is also the only file that firehol
# checks how old is it. If it is 90+ days old, firehol will complain again.
#
# Changed the supplied get-iana.sh script to generate the RESERVED_IPS file.
# FireHOL also instructs the user to use this script if the file is missing
# or is too old.
#
# Revision 1.9  2007/04/29 19:34:11  ktsaou
# *** empty log message ***
#
# Revision 1.8  2005/06/02 15:48:52  ktsaou
# Allowed 127.0.0.1 to be in RESERVED_IPS
#
# Revision 1.7  2005/05/08 23:27:23  ktsaou
# Updated RESERVED_IPS to current IANA reservations.
#
# Revision 1.6  2004/01/10 18:44:39  ktsaou
# Further optimized and reduced PRIVATE_IPS using:
# http://www.vergenet.net/linux/aggregate/
#
# The supplied get-iana.sh uses 'aggregate-flim' if it finds it in the path.
# (aggregate-flim is the name of this program when installed on Gentoo)
#
# Revision 1.5  2003/08/23 23:26:50  ktsaou
# Bug #793889:
# Change #!/bin/sh to #!/bin/bash to allow FireHOL run on systems that
# bash is not linked to /bin/sh.
#
# Revision 1.4  2002/10/27 12:44:42  ktsaou
# CVS test
#

#
# Program that downloads the IPv4 address space allocation by IANA
# and creates a list with all reserved address spaces.
#

# IPV4_ADDRESS_SPACE_URL="http://www.iana.org/assignments/ipv4-address-space"
IPV4_ADDRESS_SPACE_URL="http://www.iana.org/assignments/ipv4-address-space/ipv4-address-space.txt"

# The program will match all rows in the file which start with a number, have a slash,
# followed by another number, for which the following pattern will also match on the
# same rows
IANA_RESERVED="(RESERVED|UNALLOCATED)"

# which rows that are matched by the above, to ignore
# (i.e. not include them in RESERVED_IPS)?
#IANA_IGNORE="(Multicast|Private use|Loopback|Local Identification)"
IANA_IGNORE="Multicast"

tempfile="/tmp/iana.$$.$RANDOM"

AGGREGATE="`which aggregate-flim 2>/dev/null`"
if [ -z "${AGGREGATE}" ]
then
	AGGREGATE="`which aggregate 2>/dev/null`"
fi

if [ -z "${AGGREGATE}" ]
then
	echo >&2
	echo >&2
	echo >&2 "WARNING"
	echo >&2 "Please install 'aggregate-flim' to shrink the list of IPs."
	echo >&2
	echo >&2
fi

echo >&2
echo >&2 "Fetching IANA IPv4 Address Space, from:"
echo >&2 "${IPV4_ADDRESS_SPACE_URL}"
echo >&2

wget -O - "${IPV4_ADDRESS_SPACE_URL}"	|\
	egrep "^ *[0-9]+/[0-9]+.*${IANA_RESERVED}"	|\
	egrep -vi "${IANA_IGNORE}"			|\
	sed "s/^ \+//g"					|\
	cut -d ' ' -f 1					|\
(
	while IFS="/" read range net
	do
		# echo >&2 "$range/$net"
		
		if [ ! $net -eq 8 ]
		then
			echo >&2 "Cannot handle network masks of $net bits ($range/$net)"
			continue
		fi
		 
		first=`echo $range | cut -d '-' -f 1`
		first=`expr $first + 0`
		last=`echo $range | cut -d '-' -f 2`
		last=`expr $last + 0`
		
		x=$first
		while [ ! $x -gt $last ]
		do
			# test $x -ne 127 && echo "$x.0.0.0/$net"
			echo "$x.0.0.0/$net"
			x=$[x + 1]
		done
	done
) | \
(
	if [ ! -z "${AGGREGATE}" -a -x "${AGGREGATE}" ]
	then
		"${AGGREGATE}"
	else
		cat
	fi
) >"${tempfile}"

echo >&2 
echo >&2 
echo >&2 "FOUND THE FOLLOWING RESERVED IP RANGES:"
printf "RESERVED_IPS=\""
i=0
for x in `cat ${tempfile}`
do
	i=$[i + 1]
	printf "${x} "
done
printf "\"\n"

if [ $i -eq 0 ]
then
	echo >&2 
	echo >&2 
	echo >&2 "Failed to find reserved IPs."
	echo >&2 "Possibly the file format has been changed, or I cannot fetch the URL."
	echo >&2 
	
	rm -f ${tempfile}
	exit 1
fi
echo >&2
echo >&2
echo >&2 "Differences between the fetched list and the list installed in"
echo >&2 "/etc/sanewall/RESERVED_IPS:"

echo >&2 "# diff /etc/sanewall/RESERVED_IPS ${tempfile}"
diff /etc/sanewall/RESERVED_IPS ${tempfile}

if [ $? -eq 0 ]
then
	touch /etc/firehol/RESERVED_IPS
	echo >&2
	echo >&2 "No differences found."
	echo >&2
	
	rm -f ${tempfile}
	exit 0
fi

echo >&2 
echo >&2 
echo >&2 "Would you like to save this list to /etc/sanewall/RESERVED_IPS"
echo >&2 "so that sanewall will automatically use it from now on?"
echo >&2
while [ 1 = 1 ]
do
	printf >&2 "yes or no > "
	read x
	
	case "${x}" in
		yes)	cp -f /etc/sanewall/RESERVED_IPS /etc/sanewall/RESERVED_IPS.old 2>/dev/null
			cat "${tempfile}" >/etc/sanewall/RESERVED_IPS || exit 1
			echo >&2 "New RESERVED_IPS written to '/etc/sanewall/RESERVED_IPS'."
			break
			;;
			
		no)
			echo >&2 "Saved nothing."
			break
			;;
			
		*)	echo >&2 "Cannot understand '${x}'."
			;;
	esac
done

rm -f ${tempfile}

